File: cadwiki-nuget/cadwiki.NetUtils/VersionExtension.vb
Imports System.Runtime.CompilerServices

Public Module VersionExtension
    <Extension()>
    Public Function IncrementRevision(ByVal version As Version) As Version
        Return version.AddVersion(0, 0, 0, 1)
    End Function
    <Extension()>
    Public Function IncrementBuild(ByVal version As Version) As Version
        Return version.IncrementBuild(True)
    End Function
    <Extension()>
    Public Function IncrementBuild(ByVal version As Version, ByVal resetLowerNumbers As Boolean) As Version
        Return AddVersion(version, 0, 0, 1, If(resetLowerNumbers, -version.Revision, 0))
    End Function
    <Extension()>
    Public Function IncrementMinor(ByVal version As Version) As Version
        Return version.IncrementMinor(True)
    End Function
    <Extension()>
    Public Function IncrementMinor(ByVal version As Version, ByVal resetLowerNumbers As Boolean) As Version
        Return AddVersion(version, 0, 1, If(resetLowerNumbers, -version.Build, 0), If(resetLowerNumbers, -version.Revision, 0))
    End Function
    <Extension()>
    Public Function IncrementMajor(ByVal version As Version) As Version
        Return version.IncrementMajor(True)
    End Function
    <Extension()>
    Public Function IncrementMajor(ByVal version As Version, ByVal resetLowerNumbers As Boolean) As Version
        Return AddVersion(version, 1, If(resetLowerNumbers, -version.Minor, 0), If(resetLowerNumbers, -version.Build, 0), If(resetLowerNumbers, -version.Revision, 0))
    End Function

    <Extension()>
    Public Function AddVersion(ByVal version As Version, ByVal pAddVersion As String) As Version
        Return AddVersion(version, New Version(pAddVersion))
    End Function
    <Extension()>
    Public Function AddVersion(ByVal version As Version, ByVal pAddVersion As Version) As Version
        Return AddVersion(version, pAddVersion.Major, pAddVersion.Minor, pAddVersion.Build, pAddVersion.Revision)
    End Function
    <Extension()>
    Public Function AddVersion(ByVal version As Version, ByVal major As Integer, ByVal minor As Integer, ByVal build As Integer, ByVal revision As Integer) As Version
        Return SetVersion(version, version.Major + major, version.Minor + minor, version.Build + build, version.Revision + revision)
    End Function
    <Extension()>
    Public Function SetVersion(ByVal version As Version, ByVal major As Integer, ByVal minor As Integer, ByVal build As Integer, ByVal revision As Integer) As Version
        Return New Version(major, minor, build, revision)
    End Function

End Module

File: cadwiki-nuget/cadwiki.NetUtils/VersionExtension.cs
using System;

namespace cadwiki.NetUtils
{

    public static class VersionExtension
    {
        public static Version IncrementRevision(this Version version)
        {
            return version.AddVersion(0, 0, 0, 1);
        }
        public static Version IncrementBuild(this Version version)
        {
            return version.IncrementBuild(true);
        }
        public static Version IncrementBuild(this Version version, bool resetLowerNumbers)
        {
            return version.AddVersion(0, 0, 1, resetLowerNumbers ? -version.Revision : 0);
        }
        public static Version IncrementMinor(this Version version)
        {
            return version.IncrementMinor(true);
        }
        public static Version IncrementMinor(this Version version, bool resetLowerNumbers)
        {
            return version.AddVersion(0, 1, resetLowerNumbers ? -version.Build : 0, resetLowerNumbers ? -version.Revision : 0);
        }
        public static Version IncrementMajor(this Version version)
        {
            return version.IncrementMajor(true);
        }
        public static Version IncrementMajor(this Version version, bool resetLowerNumbers)
        {
            return version.AddVersion(1, resetLowerNumbers ? -version.Minor : 0, resetLowerNumbers ? -version.Build : 0, resetLowerNumbers ? -version.Revision : 0);
        }

        public static Version AddVersion(this Version version, string pAddVersion)
        {
            return version.AddVersion(new Version(pAddVersion));
        }
        public static Version AddVersion(this Version version, Version pAddVersion)
        {
            return version.AddVersion(pAddVersion.Major, pAddVersion.Minor, pAddVersion.Build, pAddVersion.Revision);
        }
        public static Version AddVersion(this Version version, int major, int minor, int build, int revision)
        {
            return version.SetVersion(version.Major + major, version.Minor + minor, version.Build + build, version.Revision + revision);
        }
        public static Version SetVersion(this Version version, int major, int minor, int build, int revision)
        {
            return new Version(major, minor, build, revision);
        }

    }
}